src/normalize/core.ts

```diff
@@ -55,43 +58,80 @@ export class CoreNormalizer extends SpecMapper<NormalizerParams, FacetedUnitSpec
   // This is for normalizing non-facet unit
   public mapUnit(spec: UnitSpec, params: NormalizerParams): NormalizedUnitSpec | NormalizedLayerSpec {
     const {parentEncoding, parentProjection} = params;
+
+    const specWithReplacedEncoding = {...spec, encoding: replaceRepeaterInEncoding(spec.encoding, params.repeater)};
+
     if (parentEncoding || parentProjection) {
-      return this.mapUnitWithParentEncodingOrProjection(spec, params);
+      return this.mapUnitWithParentEncodingOrProjection(specWithReplacedEncoding, params);
     }
 
     const normalizeLayerOrUnit = this.mapLayerOrUnit.bind(this);
 
     for (const unitNormalizer of this.nonFacetUnitNormalizers) {
-      if (unitNormalizer.hasMatchingType(spec, params.config)) {
-        return unitNormalizer.run(spec, params, normalizeLayerOrUnit);
+      if (unitNormalizer.hasMatchingType(specWithReplacedEncoding, params.config)) {
+        return unitNormalizer.run(specWithReplacedEncoding, params, normalizeLayerOrUnit);
       }
     }
 
-    return spec as NormalizedUnitSpec;
+    return specWithReplacedEncoding as NormalizedUnitSpec;
   }
 
-  protected mapRepeat(
-    spec: GenericRepeatSpec<UnitSpec, ExtendedLayerSpec>,
-    params: NormalizerParams
-  ): GenericRepeatSpec<NormalizedUnitSpec, NormalizedLayerSpec> {
-    const {repeat} = spec;
+  protected mapRepeat(spec: RepeatSpec, params: NormalizerParams): GenericConcatSpec<NormalizedSpec> {
+    const {repeat, spec: childSpec, data, ...remainingProperties} = spec;
 
     if (!isArray(repeat) && spec.columns) {
       // is repeat with row/column
       spec = omit(spec, ['columns']);
       log.warn(log.message.columnsNotSupportByRowCol('repeat'));
     }
 
+    const children: NormalizedSpec[] = [];
+
+    const repeater = params.repeater;
+
+    const row = (!isArray(repeat) && repeat.row) || [repeater ? repeater.row : null];
+    const column = (!isArray(repeat) && repeat.column) || [repeater ? repeater.column : null];
+    const repeatValues = (isArray(repeat) && repeat) || [repeater ? repeater.repeat : null];
+
+    // cross product
+    for (const repeatValue of repeatValues) {
+      for (const rowValue of row) {
+        for (const columnValue of column) {
+          const childRepeater = {
+            repeat: repeatValue,
+            row: rowValue,
+            column: columnValue
+          };
+
+          const child = this.map(childSpec, {...params, repeater: childRepeater});
+
+          // TODO: remove child__repeat_ prefix
+          child.name =
```

Comment:
This line is wrong, you should only include `child__repeat` only once in the beginning.  (Hence the name changes in the diff.)